hawkmoth/parser.py

```diff
@@ -255,6 +256,11 @@ def parse(filename, **options):
                      TranslationUnit.PARSE_DETAILED_PROCESSING_RECORD |
                      TranslationUnit.PARSE_SKIP_FUNCTION_BODIES)
 
+    for diag in tu.diagnostics:
+        if diag.severity >= 2:
```

Comment:
So I've tried it out and a missing header file is indeed level 4 (the highest), yet it does *not* affect the documentation. Clang does the right thing as far as I can see and just keeps going.

Maybe I didn't understand what you meant before? Or do you have an example where that occurs?

Strangely though, I added blatant syntax errors and this thing picks nothing up and documentation is still fine as long as the syntax errors don't involve blocks boundaries. It seems whatever Clang issues here is totally useless for our purposes since it doesn't matter for our task: the documentation.

I expect Clang collects those other errors in some way. I think we need to figure out how, because those are the errors that actually matter for hawkmoth. Would you agree?

Reply:
> Maybe I didn't understand what you meant before? Or do you have an example where that occurs?

In our case, the parser couldn't find the proper header files such that macros weren't expanded leading to the comment being added but not the function definition. I debugged a bit of it by creating a simple python script that that iterated over all the symbols (it literally stopped when it couldn't find the symbol). 

And for an example:

Here's what Hawmoth generates with the function_attributes branch for test/example-72-function-attrs-with-macros.c file:

```

.. c:function:: int hidden_frob(struct list * list, enum mode mode)

   List frobnicator with visibility hidden attribute.

   :param list: The list to frob.
   :param mode: The frobnication mode.
   :return: 0 on success, non-zero error code on error.
   :since: v0.1

**Visibility** hidden
```

I would then remove the macros and place them into a header file (and include that header file, but without a proper path):

Here's what it prints (just the comments):

```

List frobnicator with visibility hidden attribute.

:param list: The list to frob.
:param mode: The frobnication mode.
:return: 0 on success, non-zero error code on error.
:since: v0.1

```

>  I expect Clang collects those other errors in some way. I think we need to figure out how, because those are the errors that actually matter for hawkmoth. Would you agree?

Yes I agree. The reason which you do not see that is because of `TranslationUnit.PARSE_SKIP_FUNCTION_BODIES` passed to parse method. Once you remove that you'll also see those errors.  @jnikula do you think we should remove this such that we display compiler errors? 

I personally don't have strong opinions here. Somehow not being able to build your code takes bigger precedence than the ability to generate documentation out of it so I'd expect that code builds properly and then you generate documentation out of it. I'd argue that tools that generate documentation should take care to make sure the code is valid. IDK you decide.

Reply:
I regret I haven't been able to look into this in depth.

For things to work properly in a complex project I expect you have to pass the CFLAGS to Clang through `cautodoc_clang` config option or `clang` directive option. That's not about the project building in general, it's about running Clang with the same flags.

The question is, if you fail to do so, and you get incomplete/incorrect documentation, what are your options? I'm inclined to think we should keep the Clang build warnings to a minimum in the happy day scenario. This would mean using the Sphinx logging API, with suitable loglevels to keep the happy day scenario clean. The user can then use, say, `sphinx-build -vvv` or some number of `-v`s to get more data.

`TranslationUnit.PARSE_SKIP_FUNCTION_BODIES` is an optimization. We don't usually need anything from within the functions. If we ensure we don't rely on the state of that option in the parser, we could leave it out depending on Sphinx loglevel to gain more debug info.

Reply:
> For things to work properly in a complex project I expect you have to pass the CFLAGS to Clang through cautodoc_clang config option or clang directive option. That's not about the project building in general, it's about running Clang with the same flags.

Even with those properly added standard include files require standard clang header files (which are _not_ added by the parser). One thing that would like would be to improve hawkmoth documentation to mention this somewhere because I struggled a bit with it. 

> The question is, if you fail to do so, and you get incomplete/incorrect documentation, what are your options? I'm inclined to think we should keep the Clang build warnings to a minimum in the happy day scenario. This would mean using the Sphinx logging API, with suitable loglevels to keep the happy day scenario clean. The user can then use, say, sphinx-build -vvv or some number of -vs to get more data.

Yep, I tend to agree with this, and it seems reasonable to me.

Reply:
Ok, so I must admit you've both turned my own argument against me, and quite rightly so!

I was hoping for it to go the other way around: happy day scenario means not caring about missing stuff (including half the flags), but caring about syntax that may break the parser (e.g. bad function boundaries affecting placement of comments). Turns out both can break documentation, so my argument is out.

So yeah, I guess it does make sense to assume all flags are passed to Clang. And as long as we assume flags are being passed entirely, then level 2 seems appropriate.

We can still argue in favour of removing `TranslationUnit.PARSE_SKIP_FUNCTION_BODIES`, but it can come separately after we gain more experience with this.